File: Aoc_Kotlin/src/main/kotlin/day05/ovelapping/Day05.kt
package day05.ovelapping

import AoC_Lib.SomeDay
import AoC_Lib.Line
import AoC_Lib.Pos
import AoC_Lib.toLines

object Day05: SomeDay(2021,5) {
    override val title = "Hydrothermal Venture"

    private fun makeLines(s: String): List<Line>  =
        s.lines()
            .map { it.split(" -> ")
                .map(Pos::fromString) }
            .flatMap { it.toLines() }

    override fun first(data: String): Any? {
        val ls = makeLines(data)
        return ls.filter(Line::isVH)
            .flatMap(Line::expanded)
            .groupingBy{it}.eachCount()
            .count { it.value > 1 }
    } // 5835 Time: 115ms

    override fun second(data: String): Any? {
        val ls = makeLines(data)
        return ls
            .flatMap(Line::expanded)
            .groupingBy{it}.eachCount()
            .count { it.value > 1 }
    } // 17013 Time: 124ms
}

fun main() = SomeDay.mainify(Day05)

File: AoC_Scala3/src/day05/day05.scala
package day05

import AoC_Lib.*
import Pos.toPos
import IterableExtensions.groupCount

object day05 {
  val ls: Seq[Line] = inputStrs(day = 5)
    .map(_.split(" -> ")
      .map(_.toPos))
    .map{ case Array(p1, p2) => Line(p1,p2) }

  def overlappingPoints(ls: Seq[Line], filter: Line => Boolean = _ => true): Int =
    ls.filter(filter)
      .flatMap(_.expand)
      .groupCount(identity)
      .count(_._2 > 1)

  def part1: Int = overlappingPoints(ls, _.isVH)
  def part2: Int = overlappingPoints(ls)

  @main
  def HydrothermalVenture(): Unit = {
    Console.println(part1) // 5835
    Console.println(part2) // 17013
  }
}
